File: cpp/quick_sort_step.cpp
#include <iostream>
#include <vector>
#include <string>
#include <fstream>
#include <sstream>
#include <algorithm> 

struct DataEntry {
    int id;
    std::string value;

    bool operator<(const DataEntry& other) const {
        return id < other.id;
    }
};

void printVectorToFile(std::ofstream& outFile, const std::vector<DataEntry>& vec, int low_idx, int high_idx) {
    outFile << "[";
    for (int i = low_idx; i <= high_idx; ++i) {
        outFile << vec[i].id << "/" << vec[i].value;
        if (i < high_idx) {
            outFile << ", ";
        }
    }
    outFile << "]\n";
}

int partition_step(std::vector<DataEntry> &vec, int low, int high, std::ofstream& outFile, int original_low, int original_high) {
    DataEntry pivot = vec[high]; 
    int i = (low - 1);          

    for (int j = low; j <= high - 1; j++) {
        if (vec[j].id <= pivot.id) {
            i++;
            std::swap(vec[i], vec[j]);
        }
    }
    std::swap(vec[i + 1], vec[high]);

    outFile << "pi=" << (i + 1) - original_low << " "; 
    printVectorToFile(outFile, vec, original_low, original_high);

    return (i + 1); 
}

void quick_sort_step(std::vector<DataEntry> &vec, int low, int high, std::ofstream& outFile, int original_low, int original_high) {
    if (low < high) {
        int pi = partition_step(vec, low, high, outFile, original_low, original_high);

        quick_sort_step(vec, low, pi - 1, outFile, original_low, original_high);
        quick_sort_step(vec, pi + 1, high, outFile, original_low, original_high);
    }
}

std::vector<DataEntry> readCsvData(const std::string& filename, int start_row, int end_row) {
    std::vector<DataEntry> data;
    std::ifstream file(filename); // Open the CSV file

    if (!file.is_open()) {
        std::cerr << "Error: Cannot open file '" << filename << "'" << std::endl;
        return data; 
    }

    std::string line;
    int current_row = 0;
    while (std::getline(file, line)) {
        current_row++;
        if (current_row >= start_row && current_row <= end_row) {
            std::stringstream ss(line); 
            std::string id_str, value_str;

            if (std::getline(ss, id_str, ',') && std::getline(ss, value_str)) {
                DataEntry entry;
                try {
                    entry.id = std::stoi(id_str);
                    entry.value = value_str;
                    data.push_back(entry); 
                } catch (const std::invalid_argument& e) {
                    std::cerr << "Warning: Invalid integer format in row " << current_row << ": '" << id_str << "'" << std::endl;
                } catch (const std::out_of_range& e) {
                    std::cerr << "Warning: Integer out of range in row " << current_row << ": '" << id_str << "'" << std::endl;
                }
            }
        }
        if (current_row > end_row) {
            break;
        }
    }
    file.close();
    return data;
}

int main() {
    std::string dataset_filename;
    int start_row;
    int end_row;

    std::cout << "Enter the dataset filename (e.g., dataset_1000.csv): ";
    std::cin >> dataset_filename;

    std::cout << "Enter the starting row number: ";
    std::cin >> start_row;

    std::cout << "Enter the ending row number: ";
    std::cin >> end_row;

    if (start_row <= 0 || end_row < start_row) {
        std::cerr << "Error: Invalid row numbers. Start row must be positive and less than or equal to end row." << std::endl;
        return 1;
    }

    std::string output_filename = "quick_sort_step_" + std::to_string(start_row) + "_" + std::to_string(end_row) + ".txt";

    std::vector<DataEntry> data_to_sort = readCsvData(dataset_filename, start_row, end_row);
    if (data_to_sort.empty()) {
        std::cerr << "No data was read or file could not be opened. Exiting." << std::endl;
        return 1;
    }

    std::ofstream outFile(output_filename);
    if (!outFile.is_open()) {
        std::cerr << "Error: Cannot create output file '" << output_filename << "'" << std::endl;
        return 1;
    }

    printVectorToFile(outFile, data_to_sort, 0, data_to_sort.size() - 1);

    quick_sort_step(data_to_sort, 0, data_to_sort.size() - 1, outFile, 0, data_to_sort.size() - 1);

    outFile.close();
    std::cout << "Quick sort steps logged to " << output_filename << std::endl;

    return 0;
}

File: java/quick_sort_step.java
import java.io.*;
import java.util.*;
import java.util.stream.Collectors;

public class quick_sort_step {

    private static class Record {
        final int number;
        final String text;

        Record(int number, String text) {
            this.number = number;
            this.text = text;
        }

        @Override
        public String toString() {
            return number + "/" + text;
        }
    }

    private static class Visualizer {
        private final BufferedWriter writer;

        Visualizer(BufferedWriter writer) {
            this.writer = writer;
        }

        public void logInitialState(List<Record> records) throws IOException {
            writer.write(formatRecords(records));
            writer.newLine();
        }

        public void logPartition(int pivotIndex, List<Record> records) throws IOException {
            writer.write("pi=" + pivotIndex + " ");
            writer.write(formatRecords(records));
            writer.newLine();
        }

        private String formatRecords(List<Record> records) {
            return records.stream()
                          .map(Record::toString)
                          .collect(Collectors.joining(", ", "[", "]"));
        }
    }

    public static void main(String[] args) {
        try (Scanner sc = new Scanner(System.in)) {
            System.out.print("Enter dataset filename (e.g. dataset_1000.csv): ");
            String inputFile = sc.nextLine().trim();

            System.out.print("Enter start row: ");
            int startRow = sc.nextInt();
            System.out.print("Enter end row: ");
            int endRow = sc.nextInt();

            if (startRow <= 0 || endRow < startRow) {
                System.err.println("Error: Invalid row range. Please provide positive, sequential rows.");
                return;
            }

            String outputFile = "quick_sort_step_" + startRow + "_" + endRow + ".txt";

            try (BufferedWriter bw = new BufferedWriter(new FileWriter(outputFile))) {
                Visualizer visualizer = new Visualizer(bw);
                List<Record> subset = readCsvSubset(inputFile, startRow, endRow);

                if (subset.isEmpty()) {
                    System.err.println("Error: No data read from the specified range in the input file, or file is empty/not found.");
                    return;
                }

                visualizer.logInitialState(subset);
                quickSort(subset, 0, subset.size() - 1, visualizer);

                System.out.println("Successfully wrote visualization to: " + outputFile);
            }
        } catch (InputMismatchException e) {
            System.err.println("Error: Please enter valid integer numbers for rows. " + e.getMessage());
        } catch (FileNotFoundException e) {
            System.err.println("Error: Input file not found: " + e.getMessage());
        } catch (IOException e) {
            System.err.println("An unexpected I/O error occurred: " + e.getMessage());
        } catch (NumberFormatException e) {
            System.err.println("Error: CSV file contains invalid number format in the ID column. " + e.getMessage());
        }
    }

    public static List<Record> readCsvSubset(String filename, int startRow, int endRow) throws IOException {
        List<Record> result = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            for (int i = 1; (line = br.readLine()) != null; i++) {
                if (i >= startRow && i <= endRow) {
                    String[] parts = line.split(",", 2);
                    if (parts.length < 2) {
                        System.err.println("Warning: Skipping malformed line " + i + ": " + line);
                        continue; 
                    }
                    try {
                        result.add(new Record(Integer.parseInt(parts[0].trim()), parts[1].trim()));
                    } catch (NumberFormatException e) {
                        throw new NumberFormatException("Invalid number format in row " + i + ": " + parts[0].trim());
                    }
                }
                if (i > endRow) {
                    break;
                }
            }
        }
        return result;
    }

    public static void quickSort(List<Record> arr, int low, int high, Visualizer visualizer) throws IOException {
        if (low < high) {
            final int INSERTION_SORT_CUTOFF = 1; 
            
            if (high - low + 1 < INSERTION_SORT_CUTOFF) {
                insertionSort(arr, low, high);
            } else {
                int pi = partition(arr, low, high);
                visualizer.logPartition(pi, arr); 

                // Recursive calls for sub-arrays
                quickSort(arr, low, pi - 1, visualizer);
                quickSort(arr, pi + 1, high, visualizer);
            }
        }
    }

    public static int partition(List<Record> arr, int low, int high) {
        Record pivot = arr.get(high); 
        int i = (low - 1); 

        for (int j = low; j < high; j++) {
            if (arr.get(j).number <= pivot.number) {
                i++;
                Collections.swap(arr, i, j);
            }
        }
        Collections.swap(arr, i + 1, high); 
        return (i + 1); 
    }

    private static <T extends quick_sort_step.Record> void insertionSort(List<T> list, int low, int high) {
        for (int p = low + 1; p <= high; p++) {
            T tmp = list.get(p);
            int j;
            for (j = p; j > low && list.get(j - 1).number > tmp.number; j--) {
                list.set(j, list.get(j - 1));
            }
            list.set(j, tmp);
        }
    }
}
